spiffworkflow-backend/src/spiffworkflow_backend/scripts/delete_process_instances_with_criteria.py

```diff
@@ -0,0 +1,61 @@
+"""Delete_process_instances_with_criteria."""
+from time import time
+from typing import Any
+
+from flask_bpmn.models.db import db
+from sqlalchemy import or_
+
+from spiffworkflow_backend.models.process_instance import ProcessInstanceModel
+from spiffworkflow_backend.models.script_attributes_context import (
+    ScriptAttributesContext,
+)
+from spiffworkflow_backend.models.spiff_step_details import SpiffStepDetailsModel
+from spiffworkflow_backend.scripts.script import Script
+
+
+class DeleteProcessInstancesWithCriteria(Script):
+    """DeleteProcessInstancesWithCriteria."""
+
+    def get_description(self) -> str:
+        """Get_description."""
+        return "Delete process instances that match the provided criteria,"
+
+    def run(
+        self,
+        script_attributes_context: ScriptAttributesContext,
+        *args: Any,
+        **kwargs: Any,
+    ) -> Any:
+        """Run."""
+        criteria_list = args[0]
+
+        delete_criteria = []
+        delete_time = time()
+
+        for criteria in criteria_list:
+            delete_criteria.append(
+                (ProcessInstanceModel.process_model_identifier == criteria["name"])
+                & ProcessInstanceModel.status.in_(criteria["status"])  # type: ignore
+                & (
+                    ProcessInstanceModel.updated_at_in_seconds
+                    < (delete_time - criteria["last_updated_delta"])
+                )
+            )
+
+        results = ProcessInstanceModel.query.filter(or_(*delete_criteria)).all()
```

Comment:
i guess this will scale to like thousands but not millions, maybe? i guess if we charge $2 / instance, we'll be able to update this then. :)

Reply:
My sqlalchemy foo wasn't strong enough to just issue a delete, seemed like when I did i was below the orm and deletes didn't cascade properly. Thought about doing it all manually but then figured this would be something that always breaks when new relationships are deleted. Definitely open to suggestions on the sql part.

Also just remembered early on we talked about only doing 100 or so instances in a single pass so the system doesn't choke. Can add that in as well.

Reply:
oh, yeah, actually limiting to 100 at a time probably pretty much solves it, nice. i was also thinking along the lines of avoiding the ORM and issuing pure deletes to the db, but it probably isn't necessary with the limit, and yeah, certainly more maintainable if we can let the ORM help us with cascading.

Reply:
Ok cool, added the 100 limit so will merge - might help reproduce the db growth on dev, who knows. Also open to further improvements.